File: recursion/additive_primes.py
"""
 *****************************************************************************
   FILE:  additive_primes.py

   AUTHOR: Nathan Koike

   ASSIGNMENT: Project 7

   DATE: 16 October 2017

   DESCRIPTION: Determines if something is an additive prime number, a number
                that is prime, and one whose digits' sum is also prime
                decreasing to a single digit number that must also be prime

 *****************************************************************************
"""


def additive_prime(num):
    '''
    checks to see if a number is an additive prime, returns True or False
    '''
    
    if num < 10: # checks to see if the number does not need to have its
                 # digits summed
                 
        return prime(num)
    else:
        if prime(num): # checks to see if the number is a prime. if it is not
                       # it is impossible for that number to be an additive
                       # prime
                       
            digit_sum = sum_of_digits(num)
            
            return additive_prime(digit_sum) # loops through this function until
                                             # a boolean value is returned

        return False # if the body of the previous if statement is not run the
                     # number is automatically not an additive prime

def additive_primes_list(n):
    '''
    returns a list of all additive prime numbers from 1 to n inclusively
    '''
    
    lst = []

    for num in range(2, n + 1): # loops through all the numbers from 2 to n

        if additive_prime(num): # checks to see if the number from the iteration
                                # of the for loop is an additive prime
            
            lst.append(num) # appendss the number to the list if the number is
                            # an additive prime

    return lst
            


def prime(num):
    '''
    checks to see if a number is a prime number, returns True or False
    '''
    
    lst = [] # idea for creating a list to store remainders from a conversation
             # with Ken Fung
             
    if num > 1: # 1, 0, and negative numbers are not considered prime
        
        for x in range(2, num):
            
            remainder = num % (x) # checks to see if there is a remainder when
                                  # dividing the number by anything other than
                                  # 1 and itself
                                  
            lst.append(remainder)
            
        if 0 in lst: # there will only be a 0 in the list of remainders if the
                     # number is evenly devisible by a number other than 1 and
                     # itself
                     
            return False
        
        return True # if there is no 0 in the list of remainders the number is
                    # prime and this function returns True
                    
    return False # if the body of the if statement above is not run the number
                 # is one of the numbers not considered prime

def sum_of_digits(num):
    '''
    sums the digits of a number
    '''
    
    digit_sum = 0
    for i in range(len(str(num))): # goes through every digit in a number
        
        digit_sum += int(str(num)[i]) # adds the values of the digits together
        
    return digit_sum


def main():
    print(additive_primes_list(103))

if __name__ == "__main__":
    main()

File: recursion/additivePrimes.js
class LazyList {
  constructor(start) {
    this.head = start;
  }

  next() {
    return this.head++;
  }
}

class Seive extends LazyList {
  constructor(start) {
    super(start);

    this.found = [];
  }

  filter(value) {
    for (let prime of this.found) if (value % prime == 0) return false;
    return true;
  }

  nextPrime() {
    while (!this.filter(this.head)) this.next();

    this.found.push(this.head);

    return this.next();
  }
}

let globalSieve = new Seive(2);

function sumDigits(num) {
  let sum = 0;

  for (const digit of String(num)) sum += parseInt(digit);

  return sum;
}

function prime(num) {
  while (globalSieve.nextPrime() < num);

  return globalSieve.found.includes(num);
}

function additivePrime(num) {
  if (num < 10) {
    return prime(num);
  }

  if (prime(num)) return additivePrime(sumDigits(num));

  return false;
}

// Returns an array of all the additive prime numbers on the interval [1, n]
function additivePrimesList(n) {
  return Array.from(Array(n + 2).keys())
    .splice(2)
    .filter(additivePrime);
}

function main() {
  // console.log(additivePrimesList(10000));
  for (let i = 0; i < 10000; i++) if (additivePrime(i)) console.log(i);
}

main();
